File: packages/core/src/semantics.ts
/**
 * The M4.1 semantics wiring: derive accessibility metadata — ARIA `role`,
 * accessible `label`, active `state` flags, and the `enabled` / `selected`
 * booleans — from a widget node's props + JS bookkeeping, and push it
 * through {@link Node.setSemantics}.
 *
 * Pure bookkeeping, exactly like the native semantics store it feeds (see
 * the core `semantics` module): the descriptor is recorded on the node and
 * flushed to the native store on attach, it never reaches the scene props,
 * and the compositor never reads it — so painted cell output is
 * byte-identical whether or not semantics are wired (the store is off by
 * default, and a disabled store makes the writes inert no-ops).
 */

import type { Node } from "./index.ts";
import {
  checkboxLabels,
  radioOptions,
  selectOptions,
  toggleLabels,
} from "./index.ts";

/**
 * The accessibility semantics of a widget node — the write shape of
 * {@link Node.setSemantics}, mirroring the native `SemanticsNodeJs`:
 * `role` as its ARIA role name, the optional accessible `label`, the active
 * boolean `state` flags, and the `enabled` / `selected` booleans.
 */
export interface SemanticsDescriptor {
  /**
   * The ARIA role name: `button`, `checkbox`, `radio`, `radiogroup`,
   * `switch`, `menu`, `menuitem`, `listbox`, `textbox`, `link`, or
   * `group`.
   */
  role: string;
  /**
   * The node's accessible name (what a screen reader announces), or
   * `undefined` when the node carries no label.
   */
  label?: string;
  /**
   * The boolean semantics states currently active on the node:
   * `checked`, `focused`, `expanded`, `collapsed`. Unknown flags error
   * natively (they would silently drop otherwise).
   */
  state: string[];
  /**
   * Whether the control is interactive: `false` for a disabled /
   * read-only control.
   */
  enabled: boolean;
  /**
   * Whether the node is currently selected (a row in a listbox, an
   * option in a menu).
   */
  selected: boolean;
}

/** Whether `props[key]` is boolean-true — the defensive read of a possibly
 * undeclared widget prop (e.g. `disabled` / `focused` on a plain
 * `Input`). */
function boolProp(props: Record<string, unknown>, key: string): boolean {
  return props[key] === true;
}

/** The string label of a widget's props, or `undefined` when unset. */
function labelProp(props: Record<string, unknown>): string | undefined {
  const label = props.label;
  return typeof label === "string" ? label : undefined;
}

/** Attach `label` to a descriptor, omitting the key entirely when the label
 * is `undefined` (the project's `exactOptionalPropertyTypes` rejects an
 * explicit `label: undefined`). */
function withLabel(
  base: Omit<SemanticsDescriptor, "label">,
  label: string | undefined,
): SemanticsDescriptor {
  return label === undefined ? base : { ...base, label };
}

/**
 * Derive and record the semantics of a widget node from its current props +
 * JS bookkeeping — the M3 factory pattern: called at the end of each factory
 * (so every widget starts with semantics) and inside the rebuild functions
 * (so every state flip from `checkboxKey` / `toggleKey` / `radioKey` /
 * `selectKey` / `menuKey` — and the menu open/close path — is reflected).
 *
 * The derivation per widget kind:
 * - `checkbox` → `checkbox`, labelled from `checkboxLabels`, state
 *   `checked` + `focused`, `enabled: !disabled`.
 * - `toggle` → `switch`, labelled from `toggleLabels`, state `checked`
 *   (the native vocabulary's stand-in for the two-state `on` value) +
 *   `focused`, `enabled: !disabled`.
 * - `radio` root → `radiogroup`, state `focused` (the focused row index
 *   exists), `selected` (the group always has a selected member).
 * - `select` → `listbox`, state `expanded` (the dropdown is open) +
 *   `focused` (a highlighted row exists), `selected` (a value is
 *   confirmed / the multi selection is non-empty).
 * - `input` / `textarea` → `textbox`, labelled from props, state
 *   `focused`, `enabled: !disabled` (the native model has no `disabled`
 *   state flag — a disabled control is `enabled: false`).
 * - `menu` → `menu`, state `expanded` (the menu is open).
 *
 * A no-op for node types without a derivation (plain `text` / `box`):
 * custom semantics a consumer set on those nodes are never touched.
 */
export function syncSemantics(node: Node): void {
  const props = node.props as Record<string, unknown>;
  const focused = boolProp(props, "focused");
  switch (node.type) {
    case "checkbox": {
      const state: string[] = [];
      if (boolProp(props, "checked")) state.push("checked");
      if (focused) state.push("focused");
      node.setSemantics(withLabel({
        role: "checkbox",
        state,
        enabled: !boolProp(props, "disabled"),
        selected: false,
      }, checkboxLabels.get(node)));
      break;
    }
    case "toggle": {
      const state: string[] = [];
      if (boolProp(props, "on")) state.push("checked");
      if (focused) state.push("focused");
      node.setSemantics(withLabel({
        role: "switch",
        state,
        enabled: !boolProp(props, "disabled"),
        selected: false,
      }, toggleLabels.get(node)));
      break;
    }
    case "radio": {
      const options = radioOptions.get(node) ?? [];
      const state: string[] = [];
      const focusedIndex = typeof props.focused === "number" ? props.focused : 0;
      if (focusedIndex >= 0 && focusedIndex < options.length) {
        state.push("focused");
      }
      const selectedIndex = typeof props.selected === "number"
        ? props.selected
        : 0;
      node.setSemantics({
        role: "radiogroup",
        state,
        enabled: !boolProp(props, "disabled"),
        selected: selectedIndex >= 0 && selectedIndex < options.length,
      });
      break;
    }
    case "select": {
      const options = selectOptions.get(node) ?? [];
      const state: string[] = [];
      if (props.open === true) state.push("expanded");
      const highlighted = typeof props.highlighted === "number"
        ? props.highlighted
        : 0;
      if (highlighted >= 0 && highlighted < options.length) {
        state.push("focused");
      }
      const multi = props.multi === true;
      const value = props.value;
      const selected = multi
        ? Array.isArray(value) && value.length > 0
        : options.some((option) => option.value === value);
      node.setSemantics({
        role: "listbox",
        state,
        enabled: !boolProp(props, "disabled"),
        selected,
      });
      break;
    }
    case "input":
    case "textarea": {
      const state: string[] = [];
      if (focused) state.push("focused");
      node.setSemantics(withLabel({
        role: "textbox",
        state,
        enabled: !boolProp(props, "disabled"),
        selected: false,
      }, labelProp(props)));
      break;
    }
    case "menu": {
      const state: string[] = [];
      if (props.open === true) state.push("expanded");
      node.setSemantics({
        role: "menu",
        state,
        enabled: true,
        selected: false,
      });
      break;
    }
  }
}

File: src/core/tern-core/src/semantics.rs
//! The semantics store: accessibility-relevant metadata for scene nodes.
//!
//! A parallel map of [`SemanticsNode`] keyed by scene node id, recording what
//! each interactive node *is* (role) and *communicates* (label, state,
//! enabled, selected) for assistive technology: screen readers, a11y test
//! bridges, and the M4.2 output channel.
//!
//! The store is pure bookkeeping. It is read only through
//! [`Scene::semantics`](crate::Scene::semantics) /
//! [`Scene::semantics_iter`](crate::Scene::semantics_iter); layout and the
//! compositor never touch it, and no semantics write can change painted
//! content (semantics mutations bump the scene epoch but never push a dirty
//! id — see [`Scene::set_semantics`](crate::Scene::set_semantics)).

use std::collections::HashSet;
use std::str::FromStr;

/// The accessibility role of a scene node, following the WAI-ARIA role
/// vocabulary: each variant's [`as_str`](SemanticsRole::as_str) name is the
/// ARIA role string for it.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum SemanticsRole {
    /// An interactive push button.
    Button,
    /// A checkable toggle with a two-state value (checked / unchecked).
    Checkbox,
    /// A single selectable member of a radio group.
    Radio,
    /// A set of [`Radio`](SemanticsRole::Radio) members of which exactly one
    /// is checked.
    RadioGroup,
    /// An on/off toggle, distinct from a checkbox in that it does not submit
    /// a form value.
    Switch,
    /// A set of [`MenuItem`](SemanticsRole::MenuItem) choices.
    Menu,
    /// A single choice inside a menu.
    MenuItem,
    /// A list of selectable options, typically presented in a popup.
    Listbox,
    /// A single- or multi-line text entry field.
    Textbox,
    /// A navigation link to another location.
    Link,
    /// A generic container grouping related controls.
    Group,
}

impl SemanticsRole {
    /// The ARIA role name of this role (e.g. `"radiogroup"`, `"menuitem"`).
    pub const fn as_str(&self) -> &'static str {
        match self {
            SemanticsRole::Button => "button",
            SemanticsRole::Checkbox => "checkbox",
            SemanticsRole::Radio => "radio",
            SemanticsRole::RadioGroup => "radiogroup",
            SemanticsRole::Switch => "switch",
            SemanticsRole::Menu => "menu",
            SemanticsRole::MenuItem => "menuitem",
            SemanticsRole::Listbox => "listbox",
            SemanticsRole::Textbox => "textbox",
            SemanticsRole::Link => "link",
            SemanticsRole::Group => "group",
        }
    }
}

impl FromStr for SemanticsRole {
    type Err = ();

    /// Parse an ARIA role name (as produced by [`SemanticsRole::as_str`])
    /// into a [`SemanticsRole`]. Unknown or differently-cased names yield
    /// `Err(())`.
    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s {
            "button" => Ok(SemanticsRole::Button),
            "checkbox" => Ok(SemanticsRole::Checkbox),
            "radio" => Ok(SemanticsRole::Radio),
            "radiogroup" => Ok(SemanticsRole::RadioGroup),
            "switch" => Ok(SemanticsRole::Switch),
            "menu" => Ok(SemanticsRole::Menu),
            "menuitem" => Ok(SemanticsRole::MenuItem),
            "listbox" => Ok(SemanticsRole::Listbox),
            "textbox" => Ok(SemanticsRole::Textbox),
            "link" => Ok(SemanticsRole::Link),
            "group" => Ok(SemanticsRole::Group),
            _ => Err(()),
        }
    }
}

/// A single semantics state flag carried by a node. A node can hold several
/// at once (e.g. checked + focused); see [`SemanticsNode::state`].
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum SemanticsState {
    /// The node is checked (checkbox / radio / switch).
    Checked,
    /// The node currently has keyboard focus.
    Focused,
    /// The node's disclosure is open (menu, listbox, group).
    Expanded,
    /// The node's disclosure is closed.
    Collapsed,
}

/// The accessibility semantics of a scene node: what it is
/// ([`SemanticsNode::role`]) and what it communicates ([`SemanticsNode::label`],
/// [`SemanticsNode::state`], [`SemanticsNode::enabled`],
/// [`SemanticsNode::selected`]).
///
/// Pure data, owned by the scene's parallel semantics map; layout and the
/// compositor never read it.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct SemanticsNode {
    /// What kind of control this node is (button, checkbox, ...).
    pub role: SemanticsRole,
    /// The node's accessible name (what a screen reader announces), or `None`
    /// when the node carries no label.
    pub label: Option<String>,
    /// The boolean semantics states currently active on the node.
    pub state: HashSet<SemanticsState>,
    /// Whether the control is interactive: `false` for a disabled/read-only
    /// control.
    pub enabled: bool,
    /// Whether the node is currently selected (a row in a listbox, an option
    /// in a menu).
    pub selected: bool,
}

impl SemanticsNode {
    /// A default-enabled, unselected, unlabelled node with the given role.
    pub fn new(role: SemanticsRole) -> Self {
        Self {
            role,
            label: None,
            state: HashSet::new(),
            enabled: true,
            selected: false,
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn role_names_roundtrip_through_as_str_and_from_str() {
        // Every role serializes to its ARIA name and parses back to itself:
        // the two directions are exact inverses.
        let roles = [
            SemanticsRole::Button,
            SemanticsRole::Checkbox,
            SemanticsRole::Radio,
            SemanticsRole::RadioGroup,
            SemanticsRole::Switch,
            SemanticsRole::Menu,
            SemanticsRole::MenuItem,
            SemanticsRole::Listbox,
            SemanticsRole::Textbox,
            SemanticsRole::Link,
            SemanticsRole::Group,
        ];
        for role in roles {
            let name = role.as_str();
            assert_eq!(name.parse::<SemanticsRole>(), Ok(role), "{name:?}");
        }
    }

    #[test]
    fn role_names_match_aria_strings() {
        assert_eq!(SemanticsRole::Button.as_str(), "button");
        assert_eq!(SemanticsRole::Checkbox.as_str(), "checkbox");
        assert_eq!(SemanticsRole::Radio.as_str(), "radio");
        assert_eq!(SemanticsRole::RadioGroup.as_str(), "radiogroup");
        assert_eq!(SemanticsRole::Switch.as_str(), "switch");
        assert_eq!(SemanticsRole::Menu.as_str(), "menu");
        assert_eq!(SemanticsRole::MenuItem.as_str(), "menuitem");
        assert_eq!(SemanticsRole::Listbox.as_str(), "listbox");
        assert_eq!(SemanticsRole::Textbox.as_str(), "textbox");
        assert_eq!(SemanticsRole::Link.as_str(), "link");
        assert_eq!(SemanticsRole::Group.as_str(), "group");
    }

    #[test]
    fn unknown_and_miscased_role_names_fail_to_parse() {
        for bad in ["Button", "BUTTON", "radiogroup ", " radiogroup", "toggle", "", "menu-item"] {
            assert_eq!(bad.parse::<SemanticsRole>(), Err(()), "{bad:?} must not parse");
        }
    }

    #[test]
    fn new_node_defaults_to_enabled_unselected_unlabelled() {
        let n = SemanticsNode::new(SemanticsRole::Checkbox);
        assert_eq!(n.role, SemanticsRole::Checkbox);
        assert_eq!(n.label, None);
        assert!(n.state.is_empty());
        assert!(n.enabled);
        assert!(!n.selected);
    }

    #[test]
    fn node_carries_state_flags_and_label() {
        let mut n = SemanticsNode::new(SemanticsRole::Checkbox);
        n.label = Some("mute".to_string());
        n.state.insert(SemanticsState::Checked);
        n.state.insert(SemanticsState::Focused);
        n.selected = true;

        assert_eq!(n.label.as_deref(), Some("mute"));
        assert!(n.state.contains(&SemanticsState::Checked));
        assert!(n.state.contains(&SemanticsState::Focused));
        assert!(n.selected);

        // A disabled node is distinct from an enabled one (a11y consumers
        // announce disabled controls differently).
        assert!(n.enabled);
        n.enabled = false;
        assert!(!n.enabled);
    }
}
